### PR TITLE
Feature/ub bv 735

### PR DESCRIPTION
**Pull request by ubozov**

### CODE REVIEW

File: command/install.go [hunk 1]
```diff
@@ -39,33 +42,149 @@ func installProcessedFlags(conf *config, logger log.Logger) bool {
...
+	if err != nil {
+		return err
+	}
+	newVer := util.GetDappCtrlVersion(tempDappCtrl)
+	existVer, existPath, ok := util.ExistingDappCtrlVersion(logger)
+	if !ok || newVer > existVer {
+		conf.DappCtrl.Version = newVer
+		conf.DappCtrl.InstallPath = conf.InstallPath
+		if err := util.InstallDappCtrl(conf.DappCtrl, conf.DBEngine.DB,
+			logger, ok); err != nil {
```
Comment (drew2a) [inline command/install.go]: The wrap of the function call makes it difficult to read

File: command/install.go [hunk 2]
```diff
@@ -39,33 +42,149 @@ func installProcessedFlags(conf *config, logger log.Logger) bool {
...
+				InstallPath: existPath,
+				Service:     conf.DappCtrl.Service,
+			})
+			util.RemoveRegistryKey(conf.Registry, existVer)
+		}
+	} else {
+		util.RemoveFile(tempDappCtrl)
+	}
+	return nil
+}
```
Comment (drew2a) [inline command/install.go]: From my perspective function too long and complex. But maybe it is ok for the rest of the team.

File: util/misc_windows.go [hunk 3]
```diff
@@ -30,6 +36,12 @@ const WinRegDBEngine64 string = `SOFTWARE\PostgreSQL\Installations\`
 // WinRegDBEngine32 is a registry path to 32 bit DBEngine installations
 const WinRegDBEngine32 string = `SOFTWARE\WOW6432Node\PostgreSQL\Installations\`
 
+// WinRegInstalledDapp is a registry path to dapp installations
+const WinRegInstalledDapp string = `SOFTWARE\Privatix\Dapp\`
+
+// WinRegUninstallDapp is a registry path to dapp uninstallations
+const WinRegUninstallDapp string = `SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\Privatix Dapp `
```
Comment (drew2a) [inline util/misc_windows.go]: Maybe better to move these definitions to a config?

File: command/install.go (hunk 1 above)
Comment (ubozov) [reply command/install.go]: function was refactored

File: command/install.go (hunk 2 above)
Comment (ubozov) [reply command/install.go]: comments were added

File: util/misc_windows.go (hunk 3 above)
Comment (ubozov) [reply util/misc_windows.go]: windows registry keys have a specifictly paths for install-uninstall:
`HKEY_LOCAL_MACHINE\Software\Microsoft\Windows\CurrentVersion\Uninstall`

https://docs.microsoft.com/en-us/windows/desktop/msi/uninstall-registry-key

I think, if we move these constant values to config file we will loose control of it.

File: util/misc.go [hunk 4]
```diff
@@ -42,6 +46,24 @@ type Copy struct {
...
+// DappCtrlConfig has a config for dappctrl.
+type DappCtrlConfig struct {
+	DownloadDapp   string
+	DownloadConfig string
+	Service        *serviceConfig
+	Version        string
+	InstallPath    string
+}
+
+type serviceConfig struct {
```
Comment (furkhat) [inline util/misc.go]: it can be placed in their own package.

**Comment by ubozov** [comment]
@furkhat , codes were divided to packages. Could you review it again?

File: service/service.go [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+	"io/ioutil"
+	"os"
+	"strings"
+
+	"github.com/privatix/dapp-installer/statik"
+	"github.com/privatix/dapp-installer/util"
+)
+
+// Service has a parameters for windows service.
+type Service struct {
```
Comment (furkhat) [inline service/service.go]: This type is going to be used as `service.Service` which isn't really that pretty. Consider renaming the package. May be: `windows.Service`?

File: command/command.go [hunk 6]
```diff
@@ -26,26 +27,21 @@ Flags:
...
+	InstallPath string
+	DBEngine    *dbengine.DBEngine
+	Registry    *util.Registry
+	Dapp        *dapp.Dapp
 }
 
 func newConfig() *config {
 	return &config{
-		DBEngine: &util.DBEngine{
+		DBEngine: &dbengine.DBEngine{
```
Comment (furkhat) [inline command/command.go]: Could this be better like this `DBEngine: dbengine.NewConfig()`?

File: dapp/dapp.go [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+	DownloadCtrl   string
+	DownloadConfig string
+	DownloadGui    string
+	Service        *service.Service
+	Version        string
+	InstallPath    string
+}
+
+// GetVersion returns dapp version and downloaded path.
+func (d *Dapp) GetVersion(path string) (string, string) {
```
Comment (furkhat) [inline dapp/dapp.go]: Go convention is to omit `Get` from func name. Just `func (d *Dapp) Version(`

Comment (furkhat) [reply dapp/dapp.go]: I see you update some fields here. I think it is a bad style to set/update fields in "get" like methods. Consider renaming or decoupling.

File: util/misc.go [hunk 8]
```diff
@@ -108,41 +93,32 @@ func humanateBytes(s uint64) string {
 	return fmt.Sprintf(f, val, suffix)
 }
 
-func generateDBEngineInstallParams(dbConf *DBEngine) []string {
-	args := []string{"--mode", "unattended", "--unattendedmodeui", "none"}
-
-	if len(dbConf.ServiceName) > 0 {
-		args = append(args, "--servicename", dbConf.ServiceName)
+// GetDappCtrlVersion returns dappctrl version.
+func GetDappCtrlVersion(filename string) string {
```
Comment (furkhat) [inline util/misc.go]: Same here. Just `func DappCtrlVersion`.

File: util/win_registry.go [hunk 9]
```diff
@@ -53,3 +67,98 @@ func checkDBEngineVersion(key registry.Key, path string) (string, bool) {
...
+	if err != nil {
+		return ""
+	}
+	defer key.Close()
+
+	val, _, _ := key.GetStringValue("Common Desktop")
+	return val
+}
+
+func getProgramFilesPath() string {
```
Comment (furkhat) [inline util/win_registry.go]: same here. Just `func programFilesPath() string {`

File: command/command.go (hunk 6 above)
Comment (ubozov) [reply command/command.go]: updated

File: dapp/dapp.go (hunk 7 above)
Comment (ubozov) [reply dapp/dapp.go]: renamed

File: util/misc.go (hunk 8 above)
Comment (ubozov) [reply util/misc.go]: renamed

File: util/win_registry.go (hunk 9 above)
Comment (ubozov) [reply util/win_registry.go]: renamed

**Pull request merged**